File: src/lib/brain/retrieve-context.ts

/**
 * @fileOverview Policy-Aware Retrieval Orchestrator (Vertex-backed)
 */

import { searchArticles, searchTopics, searchInsights } from './vector-search';
import { adminDB } from '@/lib/firebase-admin';
import { rewriteQueryForRetrieval } from '@/ai/flows/rewrite-query';
import { rerankCandidates } from '@/ai/flows/rerank-candidates';

export interface AgentKnowledgePolicy {
  agentId?: string;
  isCustomerFacing: boolean;
  accessLevel:
    | 'none'
    | 'articles_only'
    | 'topics_allowed'
    | 'insights_hidden_support'
    | 'internal_full_access';
  allowedLibraryIds: string[];
}

export type AnswerMode =
  | 'article_grounded'
  | 'topic_supported'
  | 'insight_supported_hidden'
  | 'internal_evidence_only'
  | 'clarify'
  | 'escalate';

export interface RetrievalDecision {
  answerMode: AnswerMode;
  chosenCandidates: Array<{
    sourceType: 'article' | 'topic' | 'insight' | 'chunk';
    id: string;
    text: string;
    title?: string;
    url?: string;
    score: number;
  }>;
  confidence: number;
  rationale: string;
}

const CONFIDENCE_THRESHOLDS = {
  ARTICLE: 0.78,
  TOPIC: 0.75,
  INSIGHT: 0.55,
};

const BOOSTS = {
  PUBLIC: { article: 1.0, topic: 0.8, insight: 0.5 },
  INTERNAL: { article: 1.0, topic: 0.9, insight: 0.85 },
};

export async function orchestrateRetrieval(args: {
  message: string;
  hubId: string;
  spaceId: string;
  policy: AgentKnowledgePolicy;
  history?: Array<{ role: 'user' | 'assistant'; content: string }>;
  botContext?: string;
}): Promise<RetrievalDecision> {
  const { message, hubId, spaceId, policy, history, botContext } = args;

  console.log('[orchestrateRetrieval] Policy:', { accessLevel: policy.accessLevel, isCustomerFacing: policy.isCustomerFacing, agentId: policy.agentId });
  console.log('[orchestrateRetrieval] Will search insights?', policy.accessLevel === 'insights_hidden_support' || policy.accessLevel === 'internal_full_access');

  const searchQuery = await rewriteQueryForRetrieval({ query: message, history, botContext });

  const boostMap = policy.isCustomerFacing ? BOOSTS.PUBLIC : BOOSTS.INTERNAL;

  // 🧠 PARALLEL SEARCHES
  const [articles, rawTopics, insights] = await Promise.all([
    searchArticles({
      query: searchQuery,
      hubId,
      spaceId,
      allowedHelpCenterIds: policy.allowedLibraryIds,
    }),

    

    policy.accessLevel !== 'none' && policy.accessLevel !== 'articles_only'
      ? searchTopics({
          query: searchQuery,
          spaceId,
          hubId,
        })
      : Promise.resolve([]),

    policy.accessLevel === 'insights_hidden_support' ||
    policy.accessLevel === 'internal_full_access'
      ? searchInsights({
          query: searchQuery,
          hubId,
          spaceId,
        })
      : Promise.resolve([]),
  ]);

  // Expand topics: fetch actual insight content from Firestore so LLM gets real resolution steps
  const topics = await expandTopicsWithInsights(rawTopics);

  // console.log('Search results:', { articles, topics, insights });

  // 🎯 APPLY WEIGHTS
  const scoredArticles = articles.map((a) => ({
    ...a,
    weightedScore: a.score * boostMap.article,
  }));

  const scoredTopics = topics.map((t) => ({
    ...t,
    weightedScore: t.score * boostMap.topic,
  }));

  const scoredInsights = insights.map((i) => ({
    ...i,
    weightedScore: i.score * boostMap.insight,
  }));

  // 🧮 SORT
  scoredArticles.sort((a, b) => b.weightedScore - a.weightedScore);
  scoredTopics.sort((a, b) => b.weightedScore - a.weightedScore);
  scoredInsights.sort((a, b) => b.weightedScore - a.weightedScore);

  const bestArticle = scoredArticles[0];
  const bestTopic = scoredTopics[0];
  const bestInsight = scoredInsights[0];

  const topArticles = scoredArticles
    .slice(0, 5)
    .map((a) => ({ ...a, sourceType: 'article' as const }));
  const topTopics = scoredTopics
    .slice(0, 5)
    .map((t) => ({ ...t, sourceType: 'topic' as const }));
  const topInsights = scoredInsights
    .slice(0, 5)
    .map((i) => ({ ...i, sourceType: 'insight' as const }));

  const rawCandidates = [...topArticles, ...topTopics, ...topInsights];

  // 🔁 LLM RERANKING — filter irrelevant candidates and reorder by relevance
  const relevantIds = await rerankCandidates({
    query: searchQuery,
    candidates: rawCandidates,
  });

  // Rebuild ordered list: relevant first (in reranked order), then any that LLM missed (fallback)
  const rerankedMap = new Map(rawCandidates.map((c) => [c.id, c]));
  const rerankedCandidates = relevantIds
    .map((id) => rerankedMap.get(id))
    .filter((c): c is (typeof rawCandidates)[number] => !!c);

  // Use reranked if we got results, otherwise fall back to raw
  const contextCandidates = rerankedCandidates.length > 0 ? rerankedCandidates : rawCandidates;

  console.log('Retrieval context payload:', {
    articleCount: topArticles.length,
    topicCount: topTopics.length,
    insightCount: topInsights.length,
    totalCount: contextCandidates.length,
    afterRerank: rerankedCandidates.length,
  });

  console.log('=== RETRIEVAL CANDIDATES (FULL) ===');
  contextCandidates.forEach((c, i) => {
    console.log(`[${i + 1}] [${c.sourceType.toUpperCase()}] score=${c.score.toFixed(4)} title="${c.title}"`);
    console.log(`    text (${c.text.length} chars): ${c.text.substring(0, 500)}${c.text.length > 500 ? '...' : ''}`);
  });
  console.log('=== END RETRIEVAL CANDIDATES ===');

  // ============================
  // 🥇 TIER 1: ARTICLE (Highest Trust)
  // ============================
  if (bestArticle && bestArticle.score >= CONFIDENCE_THRESHOLDS.ARTICLE) {
    return {
      answerMode: 'article_grounded',
      chosenCandidates: contextCandidates,
      confidence: bestArticle.score,
      rationale: 'Strong article match found.',
    };
  }

  // ============================
  // 🥈 TIER 2: TOPIC (Pattern-based)
  // ============================
  if (bestTopic && bestTopic.score >= CONFIDENCE_THRESHOLDS.TOPIC) {
    return {
      answerMode: 'topic_supported',
      chosenCandidates: contextCandidates,
      confidence: bestTopic.score,
      rationale: 'Recurring issue pattern matched (topic).',
    };
  }

  // ============================
  // 🥉 TIER 3: INSIGHTS (Support memory)
  // ============================
  if (bestInsight && bestInsight.score >= CONFIDENCE_THRESHOLDS.INSIGHT) {
    if (policy.isCustomerFacing) {
      return {
        answerMode: 'insight_supported_hidden',
        chosenCandidates: contextCandidates,
        confidence: bestInsight.score,
        rationale: 'Internal support knowledge used (hidden from user).',
      };
    }

    return {
      answerMode: 'internal_evidence_only',
      chosenCandidates: contextCandidates,
      confidence: bestInsight.score,
      rationale: 'Internal-only insight used directly.',
    };
  }

  // ============================
  // 🆘 FALLBACK
  // ============================
  const fallbackCandidates = [
    ...scoredArticles.slice(0, 1).map((a) => ({ ...a, sourceType: 'article' as const })),
    ...scoredTopics.slice(0, 1).map((t) => ({ ...t, sourceType: 'topic' as const })),
    ...scoredInsights.slice(0, 1).map((i) => ({ ...i, sourceType: 'insight' as const })),
  ]
    .sort((a, b) => b.weightedScore - a.weightedScore)
    .slice(0, 2);

  if (fallbackCandidates.length > 0) {
    return {
      answerMode: 'clarify',
      chosenCandidates: contextCandidates,
      confidence: fallbackCandidates[0].score,
      rationale: 'Only low-confidence matches found; clarify while using best available context.',
    };
  }

  return {
    answerMode: policy.isCustomerFacing ? 'clarify' : 'escalate',
    chosenCandidates: [],
    confidence: 0,
    rationale: 'No high-confidence knowledge found.',
  };
}

type TopicResult = Awaited<ReturnType<typeof searchTopics>>[number];

/**
 * For each topic, fetch its grouped insights from Firestore and append their
 * content to the topic's text so the LLM receives actual resolution steps.
 */
async function expandTopicsWithInsights(topics: TopicResult[]): Promise<TopicResult[]> {
  if (topics.length === 0) return topics;

  return Promise.all(
    topics.map(async (topic) => {
      try {
        // Simple single-field query — no composite index needed
        const snap = await adminDB
          .collection('insights')
          .where('topicId', '==', topic.id)
          .limit(5)
          .get();

        console.log(`[expandTopics] topic="${topic.title}" → ${snap.size} insights found`);

        if (snap.empty) return topic;

        const insightTexts = snap.docs
          .map((d) => d.data().content as string | undefined)
          .filter((c): c is string => !!c && c.trim().length > 0);

        console.log(`[expandTopics] insightTexts count=${insightTexts.length}, total expanded chars=${insightTexts.reduce((s, t) => s + t.length, 0)}`);

        if (insightTexts.length === 0) return topic;

        return {
          ...topic,
          text: [topic.text, ...insightTexts].join('\n\n---\n\n'),
        };
      } catch (err) {
        console.error(`[expandTopics] Failed for topic ${topic.id}:`, err);
        return topic;
      }
    })
  );
}

File: src/lib/brain/retrieve_context.py
"""
Retrieval Orchestrator - Coordinates knowledge retrieval with fallback strategies.
"""

from typing import List, Literal, Optional
from .vector_search import search_articles, search_topics, search_insights


class AgentKnowledgePolicy:
    """Knowledge access policy for a bot."""
    def __init__(
        self,
        isCustomerFacing: bool,
        accessLevel: Literal['none', 'articles_only', 'topics_allowed', 'insights_hidden_support', 'internal_full_access'],
        allowedLibraryIds: Optional[List[str]] = None,
        agentId: Optional[str] = None
    ):
        self.agentId = agentId
        self.isCustomerFacing = isCustomerFacing
        self.accessLevel = accessLevel
        self.allowedLibraryIds = allowedLibraryIds or []


AnswerMode = Literal[
    'article_grounded',
    'topic_supported',
    'insight_supported_hidden',
    'internal_evidence_only',
    'clarify',
    'escalate'
]


class RetrievalDecision:
    """Decision result from retrieval orchestration."""
    def __init__(
        self,
        answerMode: AnswerMode,
        chosenCandidates: List[dict],
        confidence: float,
        rationale: str
    ):
        self.answerMode = answerMode
        self.chosenCandidates = chosenCandidates
        self.confidence = confidence
        self.rationale = rationale


CONFIDENCE_THRESHOLDS = {
    'ARTICLE': 0.78,
    'TOPIC': 0.75,
    'INSIGHT': 0.70,
}

BOOSTS = {
    'PUBLIC': {'article': 1.0, 'topic': 0.8, 'insight': 0.5},
    'INTERNAL': {'article': 1.0, 'topic': 0.9, 'insight': 0.85},
}


async def orchestrate_retrieval(
    message: str,
    hubId: str,
    spaceId: str,
    policy: AgentKnowledgePolicy
) -> RetrievalDecision:
    """
    Orchestrate knowledge retrieval with tiered fallback strategy.
    
    Args:
        message: User message/query
        hubId: Hub ID
        spaceId: Space ID
        policy: Knowledge access policy
        
    Returns:
        RetrievalDecision with chosen sources and answer mode
    """
    boost_map = BOOSTS['PUBLIC'] if policy.isCustomerFacing else BOOSTS['INTERNAL']
    
    # Parallel searches based on access level
    articles = await search_articles(
        query=message,
        hubId=hubId,
        spaceId=spaceId,
        allowedHelpCenterIds=policy.allowedLibraryIds if policy.accessLevel != 'none' else [],
        limit=8
    )
    
    topics = []
    if policy.accessLevel not in ['none', 'articles_only']:
        topics = await search_topics(
            query=message,
            spaceId=spaceId,
            hubId=hubId
        )
    
    insights = []
    if policy.accessLevel in ['insights_hidden_support', 'internal_full_access']:
        insights = await search_insights(
            query=message,
            hubId=hubId,
            spaceId=spaceId
        )
    
    # Apply weights
    scored_articles = [
        {**a.__dict__, 'weightedScore': a.score * boost_map['article']}
        for a in articles
    ]
    scored_topics = [
        {**t.__dict__, 'weightedScore': t.score * boost_map['topic']}
        for t in topics
    ]
    scored_insights = [
        {**i.__dict__, 'weightedScore': i.score * boost_map['insight']}
        for i in insights
    ]
    
    # Sort by weighted score
    scored_articles.sort(key=lambda x: x['weightedScore'], reverse=True)
    scored_topics.sort(key=lambda x: x['weightedScore'], reverse=True)
    scored_insights.sort(key=lambda x: x['weightedScore'], reverse=True)
    
    best_article = scored_articles[0] if scored_articles else None
    best_topic = scored_topics[0] if scored_topics else None
    best_insight = scored_insights[0] if scored_insights else None
    
    # TIER 1: ARTICLE (Highest Trust)
    if best_article and best_article['score'] >= CONFIDENCE_THRESHOLDS['ARTICLE']:
        return RetrievalDecision(
            answerMode='article_grounded',
            chosenCandidates=scored_articles[:3],
            confidence=best_article['score'],
            rationale='Strong article match found.'
        )
    
    # TIER 2: TOPIC (Pattern-based)
    if best_topic and best_topic['score'] >= CONFIDENCE_THRESHOLDS['TOPIC']:
        return RetrievalDecision(
            answerMode='topic_supported',
            chosenCandidates=[best_topic],
            confidence=best_topic['score'],
            rationale='Recurring issue pattern matched (topic).'
        )
    
    # TIER 3: INSIGHTS (Support memory)
    if best_insight and best_insight['score'] >= CONFIDENCE_THRESHOLDS['INSIGHT']:
        answer_mode = 'insight_supported_hidden' if policy.isCustomerFacing else 'internal_evidence_only'
        rationale = 'Internal support knowledge used (hidden from user).' if policy.isCustomerFacing else 'Internal-only insight used directly.'
        
        return RetrievalDecision(
            answerMode=answer_mode,
            chosenCandidates=[best_insight],
            confidence=best_insight['score'],
            rationale=rationale
        )
    
    # FALLBACK
    return RetrievalDecision(
        answerMode='clarify' if policy.isCustomerFacing else 'escalate',
        chosenCandidates=[],
        confidence=0,
        rationale='No high-confidence knowledge found.'
    )
